src/bam/simulate_reads.rs

```diff
@@ -0,0 +1,181 @@
+        {
+            record.cache_cigar();
+            //Check if mate record end within region
+            let artificial_seq = if record.is_unmapped() {
+                let mut seq = Vec::new();
+                add_random_bases(record.seq_len() as u64, &mut seq, &mut rng, &alphabet)?;
+                seq
+            } else {
+                build_sequence(
+                    &mut altered_bases,
+                    &record,
+                    (start - 1) as usize,
+                    &mut rng,
+                    &alphabet,
+                )?
+            };
+            let artificial_record = build_record(&record, &artificial_seq, (start - 1) as i64)?;
+            bam_writer.write(&artificial_record)?;
+        }
+    }
+    Ok(())
+}
+
+fn init_altered_bases(
+    original_ref: &[u8],
+    artificial_reference: &[u8],
+) -> Result<HashMap<usize, HashMap<u8, u8>>> {
+    let mut altered_bases = HashMap::new();
+    for (i, (artifical_base, original_base)) in artificial_reference
+        .iter()
+        .zip(original_ref.iter())
+        .enumerate()
+    {
+        altered_bases
+            .entry(i)
+            .or_insert_with(HashMap::new)
+            .insert(*original_base, *artifical_base);
+    }
+    Ok(altered_bases)
+}
+
+fn build_record(record: &bam::Record, artificial_seq: &[u8], offset: i64) -> Result<bam::Record> {
+    let mut artificial_record = bam::record::Record::new();
+    if let Ok(mate_cigar) = record.aux(b"MC") {
+        artificial_record.push_aux(b"MC", mate_cigar)?;
+    }
+    artificial_record.set(
+        record.qname(),
+        Some(&record.cigar()),
+        artificial_seq,
+        record.qual(),
+    );
+    artificial_record.set_pos(record.pos() - offset);
+    artificial_record.set_tid(0);
+    artificial_record.set_mtid(0);
+    artificial_record.set_mpos(record.mpos() - offset);
+    artificial_record.set_flags(record.flags());
+    artificial_record.set_insert_size(record.insert_size());
+    artificial_record.set_mapq(record.mapq());
+    Ok(artificial_record)
+}
+
+fn build_sequence(
+    altered_bases: &mut HashMap<usize, HashMap<u8, u8>>,
+    record: &bam::Record,
+    offset: usize,
+    rng: &mut ThreadRng,
+    alphabet: &[u8],
+) -> Result<Vec<u8>> {
+    let mut artificial_seq = Vec::new();
+    let record_seq = record.seq().as_bytes();
+    let mut record_pos = 0;
+    let mut ref_pos = record.pos() as usize - offset;
+    //Create random seq for leading softclips
+    for cigar in record.cigar_cached().unwrap().iter() {
+        match cigar.char() {
+            'S' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq, rng, alphabet)?;
+                record_pos += cigar.len() as usize;
+            }
+            'M' | 'X' | '=' => {
+                (0..cigar.len()).for_each(|_| {
+                    let mut reduced_alphabet = alphabet.to_vec();
+                    let base_mappings = altered_bases.get(&ref_pos).unwrap().clone();
```

Comment:
@johanneskoester I was not able to reduce the alphabet to unused bases (L145) without cloning the HashMap. Any suggestion to circumvent this?

Reply:
Yes, retain modifies the given vec. Rather use `.iter().filter(|...| ...).choose()` with the latter method call being the one from [here](https://docs.rs/rand/0.8.4/rand/seq/trait.IteratorRandom.html#method.choose).